File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        # if not root:
        #     return 0
        
        # # return (1+max(self.maxDepth(root.left),self.maxDepth(root.right))) dfs

        # stack = [[root,1]] iterative
        # res=0

        # while stack:
        #     node,v=stack.pop()

        #     if node:
        #         res = max(res,v)
        #         stack.append([node.left,v+1])
        #         stack.append([node.right,v+1])
        # return res
        q = deque() #bfs

        if root:
            q.append(root)
        level=0
        while q:
            for i in range(len(q)):
                node=q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level+=1
        return level

File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
func preorderTraversal(root *TreeNode) []int {
    res := []int{}
    stack := []*TreeNode{}
    curr := root

    for curr != nil || len(stack) > 0{
        if curr != nil{
            res = append(res, curr.Val)
            stack = append(stack, curr.Right)
            curr = curr.Left
        } else{
            curr = stack[len(stack)-1]
            stack = stack[:len(stack)-1]
        }
    }
    return res
}
